### PR TITLE
Refactor the metadata intrinsics to share code with a macro

### PR DESCRIPTION
**Pull request by hackcasual**
Support dumping debug intrinsics

Move intrinsic metadata function arguments to the CyberDWARF file. This will make WebAssembly integration easier

### CODE REVIEW

File: lib/Target/JSBackend/JSBackend.cpp [hunk 1]
```diff
@@ -3811,6 +3824,20 @@ std::string JSWriter::generateDebugRecordForVar(Metadata *MD) {
     cyberDWARFData.TypeDebugData << VarIDForJSON << ":"
     << "[6,\"" << E->getName().str() << "\"," << E->getValue() << "],";
   }
+  else if (DIExpression *E = dyn_cast<DIExpression>(MD)) {
+    // DIExrpressions are associated with metadata intrinsics
+    // They represent a sequence of DWARF operations or arguments to those operations
+    // The element values are well defined in the DWARF spec, and should be stable
+
+    cyberDWARFData.TypeDebugData << VarIDForJSON << ":"
+    << "[99";
```
Comment (kripken) [inline lib/Target/JSBackend/JSBackend.cpp]: please put this on the previous line

File: lib/Target/JSBackend/CallHandlers.h [hunk 2]
```diff
@@ -468,63 +468,42 @@ DEF_CALL_HANDLER(llvm_expect_i1, {
...
-  } else if (auto const *ValAsAssign = dyn_cast<ConstantAsMetadata>(AssignedValue)) {
-    Declares.insert("metadata_llvm_dbg_value_constant");
-    return "_metadata_llvm_dbg_value_constant(\"" + getValueAsStr(ValAsAssign->getValue())
-    + "," + VarMD + ")";
-  }
-
+std::string IntrinsicFunctionNames[2][2] = { { std::string("metadata_llvm_dbg_value_local_d"), std::string("metadata_llvm_dbg_value_local_i") },
+                                   { std::string("metadata_llvm_dbg_value_constant_d"), std::string("metadata_llvm_dbg_value_constant_i") } };
+
+#define DEF_DEBUG_METADATA_INTRINSIC( vo, lmdi, dwopi ) \
```
Comment (kripken) [inline lib/Target/JSBackend/CallHandlers.h]: could this be a normal method instead of a macro?

**Comment by kripken** [comment]
How will this help with WebAssembly?

**Comment by hackcasual** [comment]
Previously the debug intrinsics looked like this

metadata_llvm_dbg_value_local($localvar, 123, 0, 9, "localvar").

While fine JavaScript, it breaks asm.js (and WebAssembly) by not differentiating between double/int for the passed in value and using a JS string.

This change pulls the last 4 arguments into the metadata file, and emits double/int specific function calls.

**Comment by kripken** [comment]
I see, thanks.

**Comment by hackcasual** [comment]
I haven't tested it yet on WebAssembly, but it is a necessary prerequisite to getting it working

**Comment by hackcasual** [comment]
Don't merge, this breaks MSVC builds.

**Pull request closed without merging**